onedal/datatypes/data_conversion.cpp

```diff
@@ -1,318 +1,402 @@
-        const auto &homogen_input = static_cast<const dal::homogen_table &>(input);
-        if (homogen_input.get_data_layout() == dal::data_layout::row_major) {
-            const dal::data_type dtype = homogen_input.get_metadata().get_data_type(0);
-
-#define MAKE_NYMPY_FROM_HOMOGEN(NpType)                                        \
-    {                                                                          \
-        auto bytes_array = dal::detail::get_original_data(homogen_input);      \
-        res = convert_to_numpy_impl<NpType>(bytes_array,                       \
-                                            homogen_input.get_row_count(),     \
-                                            homogen_input.get_column_count()); \
-    }
-            SET_CTYPE_NPY_FROM_DAL_TYPE(
-                dtype,
-                MAKE_NYMPY_FROM_HOMOGEN,
-                throw std::invalid_argument("Not avalible to convert a numpy"));
-#undef MAKE_NYMPY_FROM_HOMOGEN
-        }
-        else {
-            throw std::invalid_argument(
-                "Output oneDAL table doesn't have row major format for homogen table");
-        }
-    }
-    else if (input.get_kind() == dal::detail::csr_table::kind()) {
-        const auto &csr_input = static_cast<const detail::csr_table &>(input);
-        const dal::data_type dtype = csr_input.get_metadata().get_data_type(0);
-#define MAKE_PY_FROM_CSR(NpType, T) \
-    { res = convert_to_py_from_csr_impl<NpType, T>(csr_input); }
-        SET_CTYPES_NPY_FROM_DAL_TYPE(
-            dtype,
-            MAKE_PY_FROM_CSR,
-            throw std::invalid_argument("Not avalible to convert a scipy.csr"));
-#undef MAKE_NYMPY_FROM_HOMOGEN
-    }
-    else {
-        throw std::invalid_argument("Output oneDAL table doesn't have homogen or csr format");
-    }
-    return res;
-}
-
-} // namespace oneapi::dal::python
+/*******************************************************************************
+* Copyright 2021 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include <stdexcept>
+#include <string>
+#include <numpy/arrayobject.h>
+
+#include "oneapi/dal/table/homogen.hpp"
+#include "oneapi/dal/table/detail/csr.hpp"
+#include "oneapi/dal/table/detail/homogen_utils.hpp"
+
+#include "onedal/datatypes/data_conversion.hpp"
+
+namespace oneapi::dal::python {
+
+#define SET_CTYPE_NPY_FROM_DAL_TYPE(_T, _FUNCT, _EXCEPTION) \
```

Comment:
Why do you move from header here?

Reply:
Because this is used just in a single `cpp` file

Reply:
I think it makes sense to move to another file, since these macros violate the readability of this file.